specs/APIReview_CustomDownload.md

```diff
@@ -0,0 +1,369 @@
+                auto showDialog = [this, args] {
+                    // Hide the default download dialog.
+                    CHECK_FAILURE(args->put_ShouldDisplayDefaultDownloadDialog(FALSE));
+
+                    UINT64 downloadSizeInBytes = 0;
+                    CHECK_FAILURE(args->get_DownloadSizeInBytes(&downloadSizeInBytes));
+
+                    wil::unique_cotaskmem_string uri;
+                    CHECK_FAILURE(args->get_Uri(&uri));
+
+                    wil::unique_cotaskmem_string mimeType;
+                    CHECK_FAILURE(args->get_MimeType(&mimeType));
+
+                    wil::unique_cotaskmem_string contentDisposition;
+                    CHECK_FAILURE(args->get_ContentDisposition(&contentDisposition));
+
+                    wil::unique_cotaskmem_string resultFilePath;
+                    CHECK_FAILURE(args->get_ResultFilePath(&resultFilePath));
+
+                    std::wstring prompt =
+                        std::wstring(L"Enter result file path or select `OK` to use default path. ") +
+                        L"Select `Cancel` to cancel the download.";
+
+                    std::wstring description = std::wstring(L"URI: ") + uri.get() + L"\r\n" +
+                                               L"Mime type: " + mimeType.get() + L"\r\n" +
+                                               L"Download size in bytes: " + std::to_wstring(downloadSizeInBytes) +
+                                               L"\r\n";
+
+                    TextInputDialog dialog(
+                        m_appWindow->GetMainWindow(), L"Download Starting", prompt.c_str(),
+                        description.c_str(), resultFilePath.get());
+                    if (dialog.confirmed)
+                    {
+                        // If user selects `OK`, the download will complete normally.
+                        // Result file path will be updated if a new one was provided.
+                        CHECK_FAILURE(args->put_ResultFilePath(dialog.input.c_str()));
+                    }
+                    else
+                    {
+                        // If user selects `Cancel`, the download will be canceled.
+                        CHECK_FAILURE(args->put_Cancel(TRUE));
+                    }
+                };
+
+                wil::com_ptr<ICoreWebView2Deferral> deferral;
+                CHECK_FAILURE(args->GetDeferral(&deferral));
+
+                m_completeDefferedDownloadEvent = [showDialog, deferral] {
+                    showDialog();
+                    CHECK_FAILURE(deferral->Complete());
+                };
+
+                return S_OK;
+            })
+            .Get(),
+        &m_DownloadStartingToken));
+```
+## .NET/ WinRT
+```c#
+webView.CoreWebView2.DownloadStarting += delegate (object sender, CoreWebView2DownloadStartingEventArgs args)
+{
+  CoreWebView2Deferral deferral = args.GetDeferral();
+  _completeDeferredDownload = new Action(() =>
```

Comment:
Can we make this sample more self contained by using 
`System.Threading.SynchronizationContext.Current.Post(theMethod, state);`
to run some code on the UI thread without needing a separate _completeDeferedDownload stuff?